### ISSUE TITLE
OpenAIModerationChain with OpenAI SDK >= v1.0.0 Broken

### ISSUE THREAD

**Issue by MrNickBreen** [langchain-ai/langchain#13685]
### System Info

It appears that OpenAI's SDK v1.0.0 update introduced some needed migrations.

Running the Langchain OpenAIModerationChain with OpenAI SDK >= v1.0.0 provides the following error:
```
You tried to access openai.Moderation, but this is no longer supported in openai>=1.0.0 - see the README at https://github.com/openai/openai-python for the API.

You can run `openai migrate` to automatically upgrade your codebase to use the 1.0.0 interface. 

Alternatively, you can pin your installation to the old version, e.g. `pip install openai==0.28`

A detailed migration guide is available here: https://github.com/openai/openai-python/discussions/742
```

After briefly reading the mitigation steps I believe the suggested migration is from `openai.Moderation.create()` -> `client.moderations.create()`. 

I believe the `validate_environment` of `OpenAIModerationChain` will want updating from ```values["client"] = openai.Moderation``` to using the recommended `client.moderations` syntax.  (https://api.python.langchain.com/en/latest/_modules/langchain/chains/moderation.html#OpenAIModerationChain) 

### Information

- [ ] The official example notebooks/scripts
- [ ] My own modified scripts

### Related Components

- [ ] LLMs/Chat Models
- [ ] Embedding Models
- [ ] Prompts / Prompt Templates / Prompt Selectors
- [ ] Output Parsers
- [ ] Document Loaders
- [ ] Vector Stores / Retrievers
- [ ] Memory
- [ ] Agents / Agent Executors
- [ ] Tools / Toolkits
- [X] Chains
- [ ] Callbacks/Tracing
- [ ] Async

### Reproduction

With OpenAI SDK >= v1.0.0 try to use `OpenAIModerationChain` to moderate a piece of content. 

Error appears:
```
You tried to access openai.Moderation, but this is no longer supported in openai>=1.0.0 - see the README at https://github.com/openai/openai-python for the API.

You can run `openai migrate` to automatically upgrade your codebase to use the 1.0.0 interface. 

Alternatively, you can pin your installation to the old version, e.g. `pip install openai==0.28`

A detailed migration guide is available here: https://github.com/openai/openai-python/discussions/742
```

### Expected behavior

When using `OpenAIModerationChain` with OpenAI SDK >= v1.0.0 I expect the chain to properly moderate content and not fail with an error.

**Comment by MrNickBreen** [comment]
dosubot I think you have the right idea, but I believe we still want to use the moderation API and not the "content filter endpoint".

**Comment by luc-kalaora** [comment]
Could you please modify the class. It will be great to use it with open-ai >= 1.0.
Suggestion:

class CustomModeration(OpenAIModerationChain):

    @root_validator()
    def validate_environment(**self**, values: Dict) -> Dict:
        """Validate that api key and python package exists in environment."""
        openai_api_key = get_from_dict_or_env(
            values, "openai_api_key", "OPENAI_API_KEY"
        )
        openai_organization = get_from_dict_or_env(
            values,
            "openai_organization",
            "OPENAI_ORGANIZATION",
            default="",
        )
        try:
            import openai

            openai.api_key = openai_api_key
            if openai_organization:
                openai.organization = openai_organization
            values["client"] = **self.client.moderations**
        except ImportError as error:
            raise ImportError(
                "Could not import openai python package. "
                "Please install it with `pip install openai`."
            ) from error
        return values

    def _call(
        self,
        inputs: Dict[str, str],
        run_manager: Optional[CallbackManagerForChainRun] = None,
    ) -> Dict[str, str]:
        text = inputs[self.input_key]
        responses = self.client.moderations.create(input=text)
        output = self._moderate(text, responses.results[0])
        return {self.output_key: output}
        
      Thanks to upgrade the code

**Comment by mrchaos** [comment]
For open-ai >= 1.0

```
 @root_validator()
    def validate_environment(cls, values: Dict) -> Dict:
        """Validate that api key and python package exists in environment."""
        openai_api_key = get_from_dict_or_env(
            values, "openai_api_key", "OPENAI_API_KEY"
        )
        openai_organization = get_from_dict_or_env(
            values,
            "openai_organization",
            "OPENAI_ORGANIZATION",
            default="",
        )
        try:
            from openai import OpenAI                                                             # <--- changed

            client = OpenAI()                                                                               # <-- changed

            values["api_key"] = client.api_key                                                      # <-- changed
            if openai_organization:
                # TODO: The 'openai.organization' option isn't read in the client API. You will need to pass it when you instantiate the client, e.g. 'OpenAI(organization=openai_organization)'
                values["openai_organization"] = openai_organization                   # <--- changed
            values["client"] = client                                                                        # <--- changed
        except ImportError:
            raise ImportError(
                "Could not import openai python package. "
                "Please install it with `pip install openai`."
            )
        return values

    @property
    def input_keys(self) -> List[str]:
        """Expect input key.

        :meta private:
        """
        return [self.input_key]

    @property
    def output_keys(self) -> List[str]:
        """Return output key.

        :meta private:
        """
        return [self.output_key]

    def _moderate(self, text: str, results: dict) -> str:
        if results.flagged:                                                                                            # <--- changed
            error_str = "Text was found that violates OpenAI's content policy."
            if self.error:
                raise ValueError(error_str)
            else:
                return error_str
        return text

    def _call(
        self,
        inputs: Dict[str, str],
        run_manager: Optional[CallbackManagerForChainRun] = None,
    ) -> Dict[str, str]:
        text = inputs[self.input_key]
        results = self.client.moderations.create(input=text)   #                           <--- changed
        output = self._moderate(text, results.results[0])   #                                 <--- changed
        return {self.output_key: output}
```

**Comment by luc-kalaora** [comment]
Thanks mrchaos for your proposition.
In validate_environment you should replace client = OpenAI() by client = OpenAI(api_key = openai_api_key)

**Comment by mattflo** [comment]
This is fixed by this PR which also includes an async implementation: https://github.com/langchain-ai/langchain/pull/18537

**Comment by ccurme** [comment]
Should be fixed in langchain 0.1.20. Closing but please let me know if the issue is not resolved and will re-open.
